cryptol-saw-core/src/Verifier/SAW/Cryptol.hs

```diff
@@ -1953,38 +2031,518 @@ exportRecordValue fields v =
+                            tl_remainder <-
+                              scGlobalApply sc "Prelude.listSortDrop"
+                                [tl_applied, n]
+                            scEithersV tl_remainder
+
+               return (typeLeft, typeRight)
+
+      let
+        scInjectRight :: Int -> Term -> IO Term
+        scInjectRight n x | n < 0     = return x
+                          | otherwise = do
+                              y <- scRight (fst (tps!!n)) (snd (tps!!n)) x
+                              scInjectRight (n-1) y
+
+        scNthInjection :: Int -> Term -> IO Term
+        scNthInjection n x = do
+                             y <- scLeft (fst (tps!!n)) (snd (tps!!n)) x
+                             scInjectRight (n-1) y
+
+      return scNthInjection
+
+  -------------------------------------------------------------
+  -- Create the definition for each constructor:
+  defn_eachCtor <-
+    flip mapM (zip argTypes_eachCtor ctors) $
+        \(argTypes, ctor)->
+        do
+        let
+          ctorName   = C.ecName ctor
+          ctorNumber = C.ecNumber ctor
+          numArgs    = length argTypes
+
+        -- NOTE: we don't add the constructor arguments to the Env, as
+        -- the only references to these arguments are in the generated
+        -- SAWCore code.
+
+        -- create the vars (& names) for constructor arguments:
+        paramVars <- reverse <$> mapM (scLocalVar sc) [0..numArgs-1]
+        let paramNames = map (\x-> Text.pack ("x" ++ show x)) [0..numArgs-1]
+
+        -- create the constructor:
+        ctorBody <- addTypeAbstractions
+                    =<< scLambdaList sc (zip paramNames argTypes)
+                    =<< scNthInjection ctorNumber
+                    =<< scTuple sc paramVars
+        checkSAWCoreTypeChecks sc (C.nameIdent ctorName) ctorBody Nothing
+        return (ctorName, ctorBody)
+
+  return defn_eachCtor
+
+
+-- | importCase - translates a Cryptol case expr to SAWCore: an application
+--   of the generated SAWCore <NAME>__case function to appropriate arguments.
+--
+--   - Note missing functionality: (FIXME)
+--     - not implemented yet: default case alternatives that bind the whole scrutinee,
+--       i.e.,
+--
+--       >  case S of
+--       >    C1 ... -> E
+--       >    y      -> E[y]  -- y == S   <- NOT IMPLEMD YET
+--
+importCase ::
+  (HasCallStack) =>
+  SharedContext -> Env ->
+  C.Type -> C.Expr -> Map C.Ident C.CaseAlt -> Maybe C.CaseAlt -> IO Term
+importCase sc env b scrutinee altsMap mDfltAlt =
+  do
+  let scrutineeTy = fastTypeOf (envC env) scrutinee
+  (nm,ctors,tyParams,tyArgs) <- case scrutineeTy of
+      C.TNominal (NominalType {ntName=nm, ntParams=tyParams, ntDef=C.Enum ctors})
+                 tyArgs ->
+          return (nm,ctors,tyParams,tyArgs)
+      _ ->
+          panic "importCase"
+            [ "`case` expression scrutinee does not have an Enum type"
+            , pretty scrutineeTy
+            ]
```

Comment:
Consider using Cryptol's [`tIsNominal`](https://github.com/GaloisInc/cryptol/blob/6653adb71146c65c8863be4f5bd0b3ae61b4ee3c/src/Cryptol/TypeCheck/Type.hs#L602-L608) function here, as it takes care of pattern-matching on the structure of the `Type` for you.